pkg/mcastmanager/mcastmanager.go

```diff
@@ -0,0 +1,136 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mcastmanager
+
+import (
+	"github.com/cilium/cilium/pkg/addressing"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/multicast"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "mcast-manager")
+)
+
+// TODO(ygui): Garbage collection
+
+// MCastManager manages IPv6 address
+type MCastManager struct {
+	mutex lock.Mutex
+
+	// ifc is the interface that mcast addresses are applied to
+	ifc string
+
+	// ipv6s keeps track of all ipv6 addresses by grouping them based on the
+	// last 3 bytes of the address
+	ipv6s map[int32]map[string]bool
+
+	// state tracks all the maddrs created by MCastManager
+	state map[string]bool
+}
+
+// New creates a McastManager instance. Create a dummy manager when ifc is empty
+// string.
+func New(ifc string) *MCastManager {
+	return &MCastManager{
+		ipv6s: make(map[int32]map[string]bool),
+		state: make(map[string]bool),
+		ifc:   ifc,
+	}
+}
+
+// AddAddress is called when a new endpoint is added
+func (mgr *MCastManager) AddAddress(ipv6 addressing.CiliumIPv6) {
+	if mgr.ifc == "" || !ipv6.IsSet() {
+		return
+	}
+
+	key := mcastKey(ipv6)
+
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	if _, ok := mgr.ipv6s[key]; !ok {
+		// First IP that has the solicited node maddr
+		mgr.joinGroup(ipv6)
+		mgr.ipv6s[key] = map[string]bool{}
+	}
+
+	mgr.ipv6s[key][ipv6.String()] = true
+}
+
+// RemoveAddress is called when an endpoint is removed
+func (mgr *MCastManager) RemoveAddress(ipv6 addressing.CiliumIPv6) {
+	if mgr.ifc == "" || !ipv6.IsSet() {
+		return
+	}
+
+	key := mcastKey(ipv6)
+
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	if m, ok := mgr.ipv6s[key]; ok {
+		delete(m, ipv6.String())
+		if len(m) == 0 {
+			// Last IP that has the solicited node maddr
+			mgr.leaveGroup(ipv6)
+			delete(mgr.ipv6s, key)
+		}
+	}
+}
+
+func (mgr *MCastManager) joinGroup(ipv6 addressing.CiliumIPv6) {
+	maddr := multicast.SolicitedNodeMaddr(ipv6.IP())
+	err := multicast.JoinGroup(mgr.ifc, maddr.String())
+	if err != nil {
+		log.Warnf("failed to join mcast group %s: %v", maddr.String(), err)
+		return
+	}
+
+	log.Infof("joined mcast group %s on NIC %s", maddr.String(), mgr.ifc)
+	mgr.state[maddr.String()] = true
+}
+
+func (mgr *MCastManager) leaveGroup(ipv6 addressing.CiliumIPv6) {
+	maddr := multicast.SolicitedNodeMaddr(ipv6.IP())
+	err := multicast.LeaveGroup(mgr.ifc, maddr.String())
+	if err != nil {
+		log.Warnf("failed to leave mcast group %s: %v", maddr.String(), err)
+		return
+	}
+
+	log.Infof("left mcast group %s on NIC %s", maddr.String(), mgr.ifc)
```

Comment:
Nit: could you capitalize the info/warning messages to be consistent with existing messages?

Also, same change with `WithField` as above is possible here.

Reply:
Ack.